common/lib/capa/capa/responsetypes.py

```diff
@@ -1351,23 +1587,163 @@ class OptionResponse(LoncapaResponse):
+
+        # If no partial credit, grade it right now.
+        if not self.has_partial_credit:
+            return self.grade_without_partial_credit(problem_map, student_answers)
+
+        # This below checks to see whether we're using an alternate grading scheme.
+        #  Set partial_credit="false" (or remove it) to require an exact answer for any credit.
+        #  Set partial_credit="points" to allow credit for listed alternative answers.
+
+        # Translators: 'partial_credit' and the items in the 'graders' object
+        # are attribute names or values and should not be translated.
+        graders = {
+            'points': self.grade_via_points,
+            'false': self.grade_without_partial_credit
+        }
+
+        # Only one type of credit at a time.
+        if len(self.credit_type) > 1:
+            raise LoncapaProblemError('Only one type of partial credit is allowed for Dropdown problems.')
+        # Make sure we're using an approved style.
+        if self.credit_type[0] not in graders:
+            raise LoncapaProblemError('partial_credit attribute should be one of: ' + ','.join(graders))
+
+        # Run the appropriate grader.
+        return graders[self.credit_type[0]](
+            problem_map=problem_map,
+            student_answers=student_answers
+        )
+
+    def get_problem_attributes(self):
+        """
+        This returns a dict built of of three smaller dictionaries.
+        Keys are:
+        "correct":
+            A dictionary with problem ids as keys.
+            Entries are lists of the correct answers for that id.
+        "partial":
+            A dictionary with problem ids as keys.
+            Entries are lists of the partially-correct answers for that id.
+        "point_values":
+            Matches the "partial" one, but gives point values instead.
+            Defaults to 50% credit.
+        """
+
+        default_credit = 0.5
+
+        problem_map = dict()
+
+        for target in ['correct', 'partial', 'point_values']:
+
+            small_map = dict([
+                (af.get('id'), contextualize_text(
+                    af.get(target, default=None),
+                    self.context
+                ))
+                for af in self.answer_fields
+            ])
+
+            for answer_id in small_map:
+                if small_map[answer_id] is not None:
+                    # Split on commas and strip whitespace
+                    # to allow for multiple options.
+                    small_map[answer_id] = small_map[answer_id].split(',')
+                    for index, word in enumerate(small_map[answer_id]):
+                        # Pick out whether we're getting numbers or strings.
+                        if target in ['point_values']:
+                            small_map[answer_id][index] = float(word.strip())
+                        else:
+                            small_map[answer_id][index] = str(word.strip())
```

Comment:
Aw, phooey. Yeah, that should work. Is there anything you need me to do on this end?

A quick look through other places where I used str() shows that they're primarily credit type names, so hopefully this won't be breaking anything else. We're just limited to ASCII for credit types.